tests/framework/ab_test.py

```diff
@@ -0,0 +1,132 @@
+this can change due to external action (a new vulnerability is discovered and published to RustSec). At this point,
+every PR would fail until this dependency is removed, blocking all development. Simply removing the test from PR CI
+is not an option, since we want to avoid the scenario where a PR adds a dependency with a known vulnerability (e.g.
+the PR itself changes the "list of vulnerable dependencies"). A/B-Testing allows us to not block PRs on the former case,
+while still preventing the latter: We run cargo audit twice, once on main HEAD, and once on the PR HEAD. If the output
+of both invocations is the same, the test passes (with us being alerted to this situtation via a special pipeline that
+does not block PRs). If not, it fails, preventing PRs from introducing new vulnerable dependencies.
+"""
+import contextlib
+import os
+from tempfile import TemporaryDirectory
+from typing import Callable, Optional, TypeVar
+
+from framework import utils
+
+# Locally, this will always compare against main, even if we try to merge into, say, a feature branch.
+# We might want to do a more sophisticated way to determine a "parent" branch here.
+DEFAULT_A_REVISION = os.environ.get("BUILDKITE_PULL_REQUEST_BASE_BRANCH", "main")
+
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def default_comparator(ah: T, be: T) -> bool:
+    """Returns `true` iff that the two arguments are equal.
+
+    The default assertion for A/B-tests using `ab_test`.
+
+    Ridiculous variable names sponsored by pylint."""
+    return ah == be
+
+
+def git_ab_test(
+    test_runner: Callable[[bool], T],
+    comparator: Callable[[T, T], U] = default_comparator,
+    *,
+    a_revision: str = DEFAULT_A_REVISION,
+    b_revision: Optional[str] = None,
+) -> (T, T, U):
+    """
+    Performs an A/B-test using the given test runner between the specified revision, and the currently checked out revision.
+    The specified revisions will be checked out in temporary directories, with `test_runner` getting executed in the
+    repository root. If the test depends on firecracker binaries built from the requested revision, care has to be taken
+    that they are built from the sources in the temporary directory (which can be ensured via the `workspace` parameter
+    to `cargo_build.get_binary`).
+
+    Note that there are no guarantees on the order in which the two tests are run.
+
+    :param test_runner: A callable which when executed runs the test in the context of the current working directory. Its
+                        parameter is `true` if and only if it is currently running the "A" test.
+    :param comparator: A callable taking two outputs from `test_runner` and comparing them. Should return some value
+                       indicating whether the test should pass or no, which will be returned by the `ab_test` functions,
+                       and on which the caller can then do an assertion.
```

Comment:
Just out of curiosity, do we have any specific usecase to return non-bool value from a comparator?

Reply:
Originally the idea was that for example a performance A/B test returns a p-value or similar. That allows the test to emit more detailed comparison output, without the comparator needing to know about the metrics component of our testing, which means they can be used more generally (say, imported into a Jupyter notebook)

Reply:
Thanks. I was just thinking of whether it needs to be generalized much.